#!/usr/bin/env python

helptext='''This script will use bwa and samtools to determine the depth of coverage across a sequence
generated by HybSeqPipeline. The user can decide to plot the coverage against only the CDS
sequence generated by Exonerate (the default) or plot against all contigs that passed
the quality thresholds during the Exonerate step.

The latter option can tell you what the coverage in the intron regions might be.
'''

import sys,os,argparse,shutil


def pileup_cds():
	bwa_index_cmd = "bwa index {}/sequences/FNA/{}.FNA".format(prefix,gene_name)
	bwa_samtools_cmd = "bwa mem -t {} {} ../../{}* | samtools view -bS - | samtools sort - cds.sorted".format(cpus,bwa_index_loc,reads_stem)
	pileup_cmd = "samtools mpileup cds.sorted.bam"


def main():
	parser = argparse.ArgumentParser(description=helptext,formatter_class=argparse.RawTextHelpFormatter)
	parser.add_argument("--intron-mode",help="Calculate coverage from all exonerate contigs, not just surviving coding sequence.",default=False,action=store_true)
	parser.add_argument("-c","--cds_fn",help="Fasta file of coding domain sequences (nucleotides), should have same names as corresponding protein file.",default=None)
	



if '__name__' == '__main__': main()